docs/src/keymap-upgrade.js

```diff
@@ -0,0 +1,234 @@
+  const tree = parser.parse(text);
+
+  const edits = [...upgradeBehaviors(tree), ...upgradeKeycodes(tree)];
+
+  return applyEdits(text, edits);
+}
+
+class TextEdit {
+  /**
+   * Creates a text edit to replace a range or node with new text.
+   * Construct with one of:
+   *
+   * * `Edit(startIndex, endIndex, newText)`
+   * * `Edit(node, newText)`
+   */
+  constructor(startIndex, endIndex, newText) {
+    if (typeof startIndex !== "number") {
+      const node = startIndex;
+      newText = endIndex;
+      startIndex = node.startIndex;
+      endIndex = node.endIndex;
+    }
+
+    /** @type number */
+    this.startIndex = startIndex;
+    /** @type number */
+    this.endIndex = endIndex;
+    /** @type string */
+    this.newText = newText;
+  }
+}
+
+/**
+ * Upgrades deprecated behavior references.
+ * @param {Parser.Tree} tree
+ */
+function upgradeBehaviors(tree) {
+  /** @type TextEdit[] */
+  let edits = [];
+
+  const query = Devicetree.query("(reference label: (identifier) @ref)");
+  const matches = query.matches(tree.rootNode);
+
+  for (const { captures } of matches) {
+    const node = findCapture("ref", captures);
+    if (node) {
+      edits.push(...getUpgradeEdits(node, Behaviors));
+    }
+  }
+
+  return edits;
+}
+
+/**
+ * Upgrades deprecated key code identifiers.
+ * @param {Parser.Tree} tree
+ */
+function upgradeKeycodes(tree) {
+  /** @type TextEdit[] */
+  let edits = [];
+
+  // No need to filter to the bindings array. The C preprocessor would have
+  // replaced identifiers anywhere, so upgrading all identifiers preserves the
+  // original behavior of the keymap (even if that behavior wasn't intended).
+  const query = Devicetree.query("(identifier) @name");
+  const matches = query.matches(tree.rootNode);
+
+  for (const { captures } of matches) {
+    const node = findCapture("name", captures);
+    if (node) {
+      edits.push(...getUpgradeEdits(node, Codes, keycodeReplaceHandler));
+    }
+  }
+
+  return edits;
+}
+
+/**
+ * @param {Parser.SyntaxNode} node
+ * @param {string | null} replacement
+ * @returns TextEdit[]
+ */
+function keycodeReplaceHandler(node, replacement) {
+  if (replacement) {
+    return [new TextEdit(node, replacement)];
+  }
+
+  const nodes = findBehaviorNodes(node);
+
+  if (nodes.length === 0) {
+    console.warn(
+      `Found deprecated code "${node.text}" but it is not a parameter to a behavior`
+    );
+    return [new TextEdit(node, `/* "${node.text}" no longer exists */`)];
+  }
+
+  const oldText = nodes.map((n) => n.text).join(" ");
+  const newText = `&none /* "${oldText}" no longer exists */`;
+
+  const startIndex = nodes[0].startIndex;
+  const endIndex = nodes[nodes.length - 1].endIndex;
+
+  return [new TextEdit(startIndex, endIndex, newText)];
+}
+
+/**
+ * Returns the node for the named capture.
+ * @param {string} name
+ * @param {any[]} captures
+ * @returns {Parser.SyntaxNode | null}
+ */
+function findCapture(name, captures) {
+  for (const c of captures) {
```

Comment:
Query API returns something that looks like

```ts
type Result = { 
  captures: { 
    name: string;
    node: Parser.SyntaxNode; 
  }[]; 
}[];
```
All the queries I'm running should just return results with a single capture, so this just finds and returns it.